java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java

```diff
@@ -0,0 +1,154 @@
+
+  private static final MethodInvocationMatcher EQUALS_MATCHER = MethodInvocationMatcher.create()
+    .name("equals")
+    .addParameter("java.lang.Object");
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    MethodTree methodTree = (MethodTree) tree;
+    if (methodTree.block() != null && "equals".equals(methodTree.symbol().name()) && methodTree.parameters().size() == 1) {
+      Symbol parameterSymbol = methodTree.parameters().get(0).symbol();
+      if (parameterSymbol.type().is("java.lang.Object")) {
+        ExpressionVisitor visitor = new ExpressionVisitor(parameterSymbol);
+        methodTree.accept(visitor);
+        if (!visitor.typeChecked) {
+          addIssue(tree, "Add a type test to this method.");
+        }
+      }
+    }
+  }
+
+  private static class ExpressionVisitor extends BaseTreeVisitor {
+    private final Symbol parameterSymbol;
+    private boolean typeChecked;
+
+    ExpressionVisitor(Symbol parameterSymbol) {
+      this.parameterSymbol = parameterSymbol;
+    }
+
+    @Override
+    public void visitInstanceOf(InstanceOfTree tree) {
+      if (isArgument(tree.expression())) {
+        typeChecked = true;
+      }
+    }
+
+    @Override
+    public void visitBinaryExpression(BinaryExpressionTree tree) {
+      if (tree.is(Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO)) {
+        if (isGetClassOnArgument(tree.leftOperand()) || isGetClassOnArgument(tree.rightOperand())) {
+          typeChecked = true;
+        } else if (isExplicitComparison(tree.leftOperand(), tree.rightOperand())) {
+          typeChecked = true;
+        }
+      } else {
+        super.visitBinaryExpression(tree);
+      }
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      if (EQUALS_MATCHER.matches(tree)) {
+        typeChecked = true;
+      }
+    }
+
+    private boolean isArgument(ExpressionTree tree) {
+      ExpressionTree expressionTree = removeParenthesis(tree);
+      return expressionTree.is(Tree.Kind.IDENTIFIER) && ((IdentifierTree) expressionTree).symbol().equals(parameterSymbol);
+    }
+
+    private boolean isExplicitComparison(ExpressionTree operand1, ExpressionTree operand2) {
+      return (isArgument(operand1) && isThis(operand2)) || (isArgument(operand2) && isThis(operand1));
+    }
+
+    private boolean isGetClassOnArgument(ExpressionTree tree) {
+      ExpressionTree expressionTree = removeParenthesis(tree);
+      if (expressionTree.is(Tree.Kind.METHOD_INVOCATION)) {
+        ExpressionTree methodSelect = ((MethodInvocationTree) expressionTree).methodSelect();
+        if (methodSelect.is(Tree.Kind.MEMBER_SELECT)) {
+          ExpressionTree expression = ((MemberSelectExpressionTree) methodSelect).expression();
+          return isArgument(expression);
+        }
+      }
+      return false;
```

Comment:
Didn't you forget to check if the invoked method is actually called "getClass" and not any other method?